### PR TITLE
[CALCITE-6727] Column uniqueness constrain should only apply to inner join

### PR DESCRIPTION
**Pull request by joeyutong**
In the existing test
```
// RelMetadataTest.java
@Test void testColumnUniquenessForJoinOnLimit1() {
  final String sql = ""
      + "select *\n"
      + "from emp A\n"
      + "join (\n"
      + "  select * from emp\n"
      + "  limit 1) B\n"
      + "on A.empno = B.empno";
  sql(sql)
      .assertThatAreColumnsUnique(bitSetOf(0), is(true))
      .assertThatAreColumnsUnique(bitSetOf(1), is(true))
      .assertThatAreColumnsUnique(bitSetOf(9), is(true))
      .assertThatAreColumnsUnique(bitSetOf(10), is(true))
      .assertThatAreColumnsUnique(bitSetOf(), is(true))
      .assertThatUniqueKeysAre(bitSetOf());
} 
```
the join result column `A.ENAME` is considered to be unique because `A.EMPNO` is the unique key. The test still passes when the query changes to left join, in which case `A.ENAME` is not guaranteed to be unique.

### CODE REVIEW

**Review by mihaibudiu — COMMENTED** [review]
Why not add the test?

**Review by suibianwanwank — APPROVED** [review]
LGTM

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (mihaibudiu) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:1196]: it is customary to add a JavaDoc comment to the test indicating the JIRA issue that is being addressed. This makes it easier for maintainers to understand the rationale for a test. You can find lots of examples in the code about how that is supposed to be structured, please follow the pattern.

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java
Comment (mihaibudiu) [inline core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java:328]: is this correct even if maxRowCount is 0?

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (joeyutong) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:1199]: Added the comment. Thanks for your reminder

File: core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java
Comment (joeyutong) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java:328]: It seems to me that the constrain still works even when the build side is empty

Comment (suibianwanwank) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java:328]: +1

Comment (joeyutong) [reply core/src/main/java/org/apache/calcite/rel/metadata/RelMdColumnUniqueness.java:328]: @mihaibudiu Any more thoughts on this? Please take another look when you have a moment

**Review by mihaibudiu — APPROVED** [review]

**Comment by ILuffZhe** [comment]
Hi @joeyutong , just a little effort and I think this PR can be merged.
We can `git commit --allow-empty` and then rebase it. I'll merge it as long as CI passes.

**Comment by ILuffZhe** [comment]
It seems Jenkins failed again. I check the logs and it's not code relevant.
Do you have any idea? @mihaibudiu

**Comment by mihaibudiu** [comment]
I think you can ignore Jenkins. It fails non deterministically

File: core/src/test/java/org/apache/calcite/test/RelMetadataTest.java
Comment (ILuffZhe) [inline core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:1204]: Sorry for asking this, should `b_empno` be unique if this case is `right join`?
If yes, do you mind adding a test?

Comment (joeyutong) [reply core/src/test/java/org/apache/calcite/test/RelMetadataTest.java:1204]: I have added the corresponding test. Thanks for you feedback.

**Pull request merged**
